### PR TITLE
Stm32h745i disco zephyr

### PR DESCRIPTION
**Pull request by GaelRap**

### CODE REVIEW

**Comment by erwango** [comment]
Can you set this PR in draft mode to avoid people trying to review as if it was an official attempt to push this ?

File: boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	aliases {
+		led0 = &green_led;
+		pwm-led0 = &red_pwm_led;
+		sw0 = &user_button;
+	};
+};
+
+&clk_hse {
+	hse-bypass;
+	clock-frequency = <DT_FREQ_M(8)>; /* STLink 8MHz clock */
```
Comment (erwango) [inline boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts:49]: According to user manual 
```suggestion
	clock-frequency = <DT_FREQ_M(25)>; /* STLink 8MHz clock */
```

File: boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	status = "okay";
+};
+
+&pll {
+	div-m = <1>;
+	mul-n = <120>;
+	div-p = <2>;
```
Comment (erwango) [inline boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts:56]:
```suggestion
	div-m = <5>;
	mul-n = <95>;
	div-p = <2>;
```

File: boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	div-q = <2>;
+	div-r = <2>;
+	clocks = <&clk_hse>;
+	status = "okay";
+};
```
Comment (erwango) [inline boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts:61]:
```suggestion
	d1cpre = <1>;
	hpre = <2>;
	d1ppre = <2>;
	d2ppre1 = <2>;
	d2ppre2 = <2>;
	d3ppre = <2>;
};
```

**Comment by erwango** [comment]
@GaelRap Can you have a try with the suggested changes ?

File: boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts (hunk 1 above)
Comment (FRASTM) [reply boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts:49]: clock-frequency = <DT_FREQ_M(25)>; /* STLink 25MHz clock */

**Comment by GaelRap** [comment]
I add the changes you suggest and I get this error at compilation
```
'st,prescaler' is marked as deprecated in 'properties:' in /home/gael/Documents/S2OPC/Zephyr/zephyr/dts/bindings/pwm/st,stm32-pwm.yaml for node /soc/timers@40001800/pwm.
devicetree error: 'd1cpre' appears in /clocks/pll@0 in /home/gael/Documents/S2OPC/Zephyr/zephyr/samples/basic/blinky/build/zephyr/zephyr.dts.pre, but is not declared in 'properties:' in /home/gael/Documents/S2OPC/Zephyr/zephyr/dts/bindings/clock/st,stm32h7-pll-clock.yaml
CMake Error at /home/gael/Documents/S2OPC/Zephyr/zephyr/cmake/modules/dts.cmake:212 (message):
  gen_defines.py failed with return code: 1
Call Stack (most recent call first):
  /home/gael/Documents/S2OPC/Zephyr/zephyr/cmake/modules/zephyr_default.cmake:121 (include)
  /home/gael/Documents/S2OPC/Zephyr/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:51 (include)
  /home/gael/Documents/S2OPC/Zephyr/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:76 (include_boilerplate)
  CMakeLists.txt:4 (find_package)


-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: /home/gael/.local/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/home/gael/Documents/S2OPC/Zephyr/zephyr/samples/basic/blinky/build -S/home/gael/Documents/S2OPC/Zephyr/zephyr/samples/basic/blinky -GNinja -DBOARD=stm32h745i_disco_m7
```

**Comment by GaelRap** [comment]
Then I modify the file `dts/bindings/clock/st,stm32h7-pll-clock.yaml` to fit with the new objects add and I get this error 
```
[112/149] Building C object zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj
FAILED: zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj 
ccache /home/gael/zephyr-sdk-0.14.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DCORE_CM7 -DHSE_VALUE=25000000 -DKERNEL -DSTM32H745xx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/gael/Documents/S2OPC/Zephyr/zephyr/include -I/home/gael/Documents/S2OPC/Zephyr/zephyr/samples/basic/blinky/build/zephyr/include/generated -I/home/gael/Documents/S2OPC/Zephyr/zephyr/soc/arm/st_stm32/stm32h7 -I/home/gael/Documents/S2OPC/Zephyr/zephyr/drivers -I/home/gael/Documents/S2OPC/Zephyr/zephyr/soc/arm/st_stm32/common -I/home/gael/Documents/S2OPC/Zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/home/gael/Documents/S2OPC/Zephyr/modules/hal/atmel/include -I/home/gael/Documents/S2OPC/Zephyr/modules/hal/stm32/stm32cube/stm32h7xx/soc -I/home/gael/Documents/S2OPC/Zephyr/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include -I/home/gael/Documents/S2OPC/Zephyr/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include/Legacy -I/home/gael/Documents/S2OPC/Zephyr/modules/hal/stm32/stm32cube/common_ll/include -isystem /home/gael/Documents/S2OPC/Zephyr/zephyr/lib/libc/minimal/include -isystem /home/gael/zephyr-sdk-0.14.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include -isystem /home/gael/zephyr-sdk-0.14.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/include-fixed -Os -imacros /home/gael/Documents/S2OPC/Zephyr/zephyr/samples/basic/blinky/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m7 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /home/gael/Documents/S2OPC/Zephyr/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/gael/Documents/S2OPC/Zephyr/zephyr/samples/basic/blinky=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/gael/Documents/S2OPC/Zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/gael/Documents/S2OPC/Zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj -MF zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj.d -o zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj -c /home/gael/Documents/S2OPC/Zephyr/zephyr/drivers/clock_control/clock_stm32_ll_h7.c
/home/gael/Documents/S2OPC/Zephyr/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:137:2: error: #error "SYS clock frequency for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
  137 | #error "SYS clock frequency for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
      |  ^~~~~
[121/149] Building C object zephyr/kernel/CMakeFiles/kernel.dir/mailbox.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /home/gael/.local/bin/cmake --build /home/gael/Documents/S2OPC/Zephyr/zephyr/samples/basic/blinky/build
```

**Comment by GaelRap** [comment]
Thank you, doing this I don't get the first error but still have the second one

**Comment by erwango** [comment]
> Thank you, doing this I don't get the first error but still have the second one

Ok, seems the CubeMX output was not quite right. Have a try with :
```
&clk_hse {
	hse-bypass;
	clock-frequency = <DT_FREQ_M(25)>; /* STLink 8MHz clock */
	status = "okay";
};

&pll {
	div-m = <5>;
	mul-n = <192>;
	div-p = <2>;
	div-q = <2>;
	div-r = <2>;
	clocks = <&clk_hse>;
	status = "okay";
};

&rcc {
	clocks = <&pll>;
	clock-frequency = <DT_FREQ_M(480)>;
	d1cpre = <1>;
	hpre = <2>;
	d1ppre = <2>;
	d2ppre1 = <2>;
	d2ppre2 = <2>;
	d3ppre = <2>;
};

```

File: boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts (hunk 2 above)
Comment (FRASTM) [reply boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts:56]: to be reviewed from HSE @ 25MHz, this set of pll properties does not give sysclock of 480MHz

File: boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+&rcc {
+	clocks = <&pll>;
+	clock-frequency = <DT_FREQ_M(480)>;
+};
+
+&usart3 {
+	pinctrl-0 = <&usart3_tx_pd8 &usart3_rx_pd9>;
+	current-speed = <115200>;
```
[the remark this answers is not in the document]
Comment (GaelRap) [reply boards/arm/stm32h745i_disco/stm32h745i_disco_m7.dts:70]: I have the same understanding as you according to https://www.st.com/content/ccc/resource/technical/layouts_and_diagrams/schematic_pack/group1/e3/ee/b5/14/fe/97/43/3d/MB1381-H745XI-B02_Schematic/files/MB1381-H745XI-B02_Schematic.pdf/jcr:content/translations/en.MB1381-H745XI-B02_Schematic.pdf
PD8 and PD9 are connected to SDRAM D13/D14 and USART1_TX and USART1_RX on the arduino connector (CN6) D1 and D0

**Comment by GaelRap** [comment]
So I tried with the new dts files I have no more errors compiling blinky project but after flashing the binaries the led still not blink
What I should look now ? the gpio of the led are well configured according to https://www.st.com/resource/en/user_manual/um2488-discovery-kits-with-stm32h745xi-and-stm32h750xb-microcontrollers-stmicroelectronics.pdf page 26

File: boards/arm/stm32h745i_disco/stm32h745i_disco_m7_defconfig [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_SOC_SERIES_STM32H7X=y
+CONFIG_SOC_STM32H745XX=y
+
+# Board config should be specified since there are 2 possible targets
+CONFIG_BOARD_STM32H745I_DISCO_M4=y
```
Comment (erwango) [inline boards/arm/stm32h745i_disco/stm32h745i_disco_m7_defconfig:7]:
```suggestion
CONFIG_BOARD_STM32H745I_DISCO_M7=y
```

**Comment by erwango** [comment]
I had a closer look on board user manual and it seems that we can't use STLink MCO output as HSE on this board, but X1 ( 25 MHz oscillator) is available and can be used.
So, additionally to the change I mentionned yesterday, you should switch HSE node configuration to:
```
&clk_hse {
	clock-frequency = <DT_FREQ_M(25)>; /* X1 */
	status = "okay";
};
```

File: boards/arm/stm32h745i_disco/board.cmake [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# STM32H745I-Disco board configuration
+# Copyright (c) 2022 Systerel (https://www.systerel.fr)
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(jlink "--device=STM32H745XI" "--speed=4000")
+
```
Comment (erwango) [inline boards/arm/stm32h745i_disco/board.cmake:6]: I managed to get the blinky working with the modifications mentioned previously and then using cube programmer as a west flash runner:
```
board_runner_args(stm32cubeprogrammer "--erase" "--port=swd" "--reset=hw")
include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
```

**Comment by GaelRap** [comment]
Thank you very much I also get the blinky working !

**Comment by erwango** [comment]
@GaelRap Now that the main part os done, please let us know if you want to continue this PR or close it and maybe come back later to start the upstream process.

**Comment by GaelRap** [comment]
I think that we can close it for now and I will continue adding some stuff by my side thanks for your help

**Comment by erwango** [comment]
> when I try to flash the sample/net/eth_native_posix project into the board it doesn't boot.

`sample/net/eth_native_posix` is meant to be used on native_posix board:
cf https://docs.zephyrproject.org/2.6.0/samples/net/eth_native_posix/README.html

I would advise to use another sample, like `samples/net/dhcpv4_client` to exercise eth API.

> Did I forget to configure a needed peripheral or my &mac peripheral is badly configured ?

Configuration looks fine.

> Also I have seen that you used STMCubeMX to generate the .dts files where could I find information doing this ?

Instructions are provided here: https://github.com/zephyrproject-rtos/hal_stm32#dtsi-files
But we've just made an update, so if you're using top of the hal_stm32 repo, you should be fine

**Comment by GaelRap** [comment]
I have a try with stm32h747i_disco_m7 with the project `samples/net/dhcpv4_client` it goes well, when I launch a dhcp server the board receive an IP address. But with stm32h745i_disco_m7 the board boot and the application seems to be well launch but it doesn't reach to get an IP address. Don't understand where the problem may come from.

Thanks for the link.

**Comment by erwango** [comment]
> But with stm32h745i_disco_m7 the board boot and the application seems to be well launch but it doesn't reach to get an IP address. Don't understand where the problem may come from.

There shouldn't be much differences between those to boards. I assume you double checked the pin configuration?
Have you been able to run an ethernet sample using STM32Cube on this board ?

**Comment by GaelRap** [comment]
I double check the pin configuration everything seems to be at the good place, I'm looking for some ethernet sample to run and I come back to you.

**Comment by GaelRap** [comment]
I didn't find out a sample totally useable but I follow this tutorial and I succeed to ping the board by ethernet https://www.youtube.com/watch?v=8r8w6mgSn1A&list=PLfIJKC1ud8gjoY2McCCqkxXWuwiC2iSap&index=8
On another hand, I also find that the ethernet module of this board work with a MII interface and not a RMII so it need to have every pin set. https://www.st.com/resource/en/user_manual/um2488-discovery-kits-with-stm32h745xi-and-stm32h750xb-microcontrollers-stmicroelectronics.pdf chapter 6.10 Ethernet.
I set every pin double check it, but the sample `dhcpv4` still doesn't work. Is it possible that the problem come through cache management ?

**Comment by GaelRap** [comment]
I have the following error 
```
devicetree error: /chosen: undefined node label 'dtcm'
```

**Comment by GaelRap** [comment]
It's better I don't have anymore the error but the project still doesn't work

**Comment by erwango** [comment]
@GaelRap I've just had a test and this doesn't make any difference on a stm32_h747i_disco board (with or w/o dtcm), So I guess you should search in another direction

**Pull request closed without merging**